### ISSUE TITLE
Implement hot reload for mocha test runner

### ISSUE THREAD

**Issue by nicojs** [stryker-mutator/stryker-js#2413]
As discussed with  @Lakitna  in #1514

> **Hot reload**. Test runners won't have to clear their test suite between tests (there is a caveat here regarding static code, see hurdles).

We want to implement this feature in the mocha test runner, which supports hot reload since ~7.2 (see https://github.com/mochajs/mocha/pull/4234). 

I want to make it optional, since it might break some test suites that don't cleanup nicely between test runs.

For example:

```js
describe('foo', () => {
  const myStub = sinon.stub(); // Bad practice, will be used across runs
  it('should work', () => {
    myStub();
    sinon.assert.calledOnce(spy)
  });
});
```

This test will succeed the first time, but fail the second time. This would mark all mutants as killed.

Suggestions:

1. `"hotReloadEnabled": true` 
    `false` is default.`
2.  `"hotReload": true` 
    `false` is default`
3.  `"reloadMode": "hot"`
    `"reload"` by default.  Leaves room for adding other modes in the future (for example "restart", which restarts the test runner process between each run)

**Comment by Lakitna** [comment]
What you're describing is basically an auto-configurator.

I think it makes sense, as it would make it really straightforward to setup Stryker in the most efficient way possible. Right now that takes some research for non-trivial projects. It would be especially useful if you use it to tell people why settings do not work for them, what the impact is, and what they have to do to fix it. It would basically allow anyone to troubleshoot most things themselves without much effort. Example: `Disabling hot reload because it seems some of your tests have side effects that do not get cleaned up properly. This comes with a performance hit. [potentially link to some (third party) docs on how to find side effects]`

I imagine doing this directly after `stryker init`.

I imagine it'll be quite a bit of work to get this done, but it's also something that can be build up over time. Maybe starting with a single configuration thing like `reloadMode`/`hotReload` and then building out to other things like `concurrency`.

I do think it will take a bunch of work to get it working, but I also think it would make the barrier of entry lower. That being said, this is not something that should get focus before 4.x has been somewhat polished.

With this in mind, I really don't mind hot-reload being off by default. And even if this doesn't get built, the default can always be changed as you said.

**Comment by nicojs** [comment]
We've discussed this in the June 24'th community meetup and decided to _not make this behavior configurable_. Instead, the test runner plugin itself should decide whether or not to reload the environment (true for static mutants). See #2922 for more information.

**Comment by nicojs** [comment]
Closed with #3393, but won't be released before StrykerJS v6 because of breaking changes.
